### DISCUSSION TITLE
Making file field as Optional

### DISCUSSION THREAD

**Discussion by ShubhamMainali** [fastapi/fastapi#9114] [Questions] [answered]
I've a form-data request for user signUp process. 
The user can or can't upload it's picture it's optional. 
I've created a api, with field types as form and for image as Uploadfile everything works perfect but it is making the image field as required. 
Is it possible to make the file input as optional.. 
Do share any reference docs..

**Comment by ycd** [comment]
Can you add a minimal reproducible example? It's hard to help.

**Comment by ihakh** [comment]
you could add like `file: bytes =File(None)` or `file: bytes = File(b'')` and check if its not none or empty like `if file is not None:` or `if file:`

**Comment by ihakh** [comment, marked as answer]
when you add a default value it became optional

**Comment by ShubhamMainali** [comment]
> Can you add a minimal reproducible example? It's hard to help.

async def send_mail(sender:EmailStr=Form(...), recipient:List[EmailStr]=Form(...),content:str=Form(...),attachments:List[UploadFile]= File(...)):

I want to make attachments as optional

**Comment by ShubhamMainali** [comment]
> you could add like `file: bytes =File(None)` or `file: bytes = File(b'')` and check if its not none or empty like `if file is not None:` or `if file:`
async def send_mail(sender:EmailStr=Form(...), recipient:List[EmailStr]=Form(...),content:str=Form(...),attachments:List[UploadFile]= File(...)):
I need to use UploadFile as I've send it to attachments and when I replace it with  attachments:List[UploadFile]= File(None)
and run function without selecting file It gives me below error 
{
  "detail": [
    {
      "loc": [
        "body",
        "attachments",
        0
      ],
      "msg": "Expected UploadFile, received: <class 'str'>",
      "type": "value_error"
    }
  ]
}

**Comment by ShubhamMainali** [comment]
Need help here on urgent basis..
Below is what I'm trying
async def send_mail(
sender:EmailStr=Form(...),recipient:List[EmailStr]=Form(...),content:str=Form(...),attachments:List[UploadFile]= File(...)
):
I'm using UploadFile as I've send to send file in attachments , this can be optional , so  when I replace it with attachments:List[UploadFile]= File(None), the docs makes it a optional field but if I didn't  select file It gives me below error
{
"detail": [
{
"loc": [
"body",
"attachments",
0
],
"msg": "Expected UploadFile, received: <class 'str'>",
"type": "value_error"
}
]
}

Any solutions or alternative will be helpful here..

**Comment by ihakh** [comment]
what about this?:
```python
async def send_mail(
sender:EmailStr=Form(...),recipient:List[EmailStr]=Form(...),content:str=Form(...),attachments:List[UploadFile]= File([])
):
```
if it is a list just set an empty list as default

**Comment by ShubhamMainali** [comment]
> what about this?:
> 
> ```python
> async def send_mail(
> sender:EmailStr=Form(...),recipient:List[EmailStr]=Form(...),content:str=Form(...),attachments:List[UploadFile]= File([])
> ):
> ```
> 
> if it is a list just set an empty list as default

If I didn't use uploadfile it would not give me a option to select file on doks then.. 
I didn't got you fully, can u explain in more detail.

**Comment by MahirMahbub** [comment]
> what about this?:
> 
> ```python
> async def send_mail(
> sender:EmailStr=Form(...),recipient:List[EmailStr]=Form(...),content:str=Form(...),attachments:List[UploadFile]= File([])
> ):
> ```
> 
> if it is a list just set an empty list as default

I does not work. Any solution?

**Comment by sumandas0** [comment]
> you could add like `file: bytes =File(None)` or `file: bytes = File(b'')` and check if its not none or empty like `if file is not None:` or `if file:`

It solved my problem

**Comment by masadaqmk** [comment]
On my End is working `icon: UploadFile = File(None)`
`    if icon:
        if icon.filename:
            pass
`

**Comment by AKASH-SAKPAL** [comment]
I am getting error when I tried to set multiple upload files optional but when I upload file/files then it’ll upload successfully.

**Need to do -** Upload multiple files when required but it should be optional.

**Code:**
```
..
@query_router.post("/uploadfiles/")
async def create_upload_files(files: Union[List[UploadFile], None] = None):
    if not files:
        return {"message": 'Files not attached'}
    else:
        return {"filenames": [file.filename for file in files]}
...
```
Tried to make multiple attachments optional using above code but it’s giving following error:
**Error:**
```
{
"detail": [
{
"loc": [
"body",
"files",
0
],
"msg": "Expected UploadFile, received: <class 'str'>",
"type": "value_error"
}
]
}
```

**Comment by JarroVGIT** [comment]
Basically you should not use either `None` or `List[UploadFile], but use an empty list as default value. Try this:

```python
@ query_router.post("/uploadfiles/")
async def create_upload_files(files: list[UploadFile] = []):
    if not files:
        return {"message": "Files not attached"}
    else:
        return {"filenames": [file.filename for file in files]}

```

**Comment by Zansi-V** [comment]
solution :-  Optional[list[UploadFile]] = None

**Comment by AKASH-SAKPAL** [comment]
> Basically you should not use either `None` or `List[UploadFile], but use an empty list as default value. Try this:
> 
> ```python
> @ query_router.post("/uploadfiles/")
> async def create_upload_files(files: list[UploadFile] = []):
>     if not files:
>         return {"message": "Files not attached"}
>     else:
>         return {"filenames": [file.filename for file in files]}
> ```

Thanks for reply.
I have run the suggested code:
```
@ query_router.post("/uploadfiles/")
async def create_upload_files(files: list[UploadFile] = []):
    if not files:
        return {"message": "Files not attached"}
    else:
        return {"filenames": [file.filename for file in files]}
```
After run code, I got following error:
```
{
  "detail": [
    {
      "loc": [
        "body",
        "files",
        0
      ],
      "msg": "Expected UploadFile, received: <class 'str'>",
      "type": "value_error"
    }
  ]
}
```

**Comment by AKASH-SAKPAL** [comment]
> solution :- Optional[list[UploadFile]] = None

Thanks for reply.
I have run the following code:
```
@ query_router.post("/uploadfiles/")
async def create_upload_files(files: Optional[list[UploadFile]] = None):
    if not files:
        return {"message": "Files not attached"}
    else:
        return {"filenames": [file.filename for file in files]}
```
After run code, I got following error:
```
{
  "detail": [
    {
      "loc": [
        "body",
        "files",
        0
      ],
      "msg": "Expected UploadFile, received: <class 'str'>",
      "type": "value_error"
    }
  ]
}
```

**Comment by JarroVGIT** [comment]
How are you calling the endpoint? It works on my end using curl.

**Comment by AKASH-SAKPAL** [comment]
> How are you calling the endpoint? It works on my end using curl.

I'm calling endpoint through Swagger UI (http://127.0.0.1:8000/docs).

**Comment by JarroVGIT** [comment]
What is the generated curl that swagger gives you?

**Comment by AKASH-SAKPAL** [comment]
> What is the generated curl that swagger gives you?

Generated curl is as follows:
```
curl -X 'POST' \
  'http://127.0.0.1:8000/student_query/uploadfiles/' \
  -H 'accept: application/json' \
  -H 'Content-Type: multipart/form-data' \
  -F 'files='
```

**Comment by JarroVGIT** [comment]
This is a swagger ui bug; rather than sending no files, it tries to send “something” as files.

**Comment by AKASH-SAKPAL** [comment]
> This is a swagger ui bug; rather than sending no files, it tries to send “something” as files.

Any other solution?

**Comment by JarroVGIT** [comment]
Is it super required that you use Swagger? I mean, the API works as intended, it is just the medium that you use to aproach it that doesn't work. Using `curl -X 'POST' 'http://127.0.0.1:8000/student_query/uploadfiles/'` from a command line will work. If you want Swagger to work properly, then I would suggest opening an issue on that project.

**Comment by AKASH-SAKPAL** [comment]
> ning an issue on that project

No super required but for testing I'm using it.
Thank you very much.
